docs/gallery/advanced_io/streaming.py

```diff
@@ -1,98 +1,119 @@
+first need to uninstall a currently installed version of ``h5py``.
+
+.. code-block:: bash
+
+   pip uninstall h5py
+   conda install -c conda-forge "h5py>=3.2"
+
+Now instantiate a :py:class:`~pynwb.NWBHDF5IO` object with the S3 URL and specify the driver as "ros3". This
+will download metadata about the file from the S3 bucket to memory. The values of datasets are accessed lazily,
+just like when reading an NWB file stored locally. So, slicing into a dataset will require additional time to
+download the sliced data (and only the sliced data) to memory.
+
+.. code-block:: python
+
+   from pynwb import NWBHDF5IO
 
-####################
-# .. code-block:: bash
-#
-#   pip uninstall h5py
-#   conda install -c conda-forge "h5py>=3.2"
-#
-
-####################
-# The ``DandiAPIClient`` can be used to get the S3 URL to an NWB file of interest stored in the DANDI Archive.
-# If you have not already, install the latest release of the ``dandi`` package.
-#
-# .. code-block:: bash
-#
-#   pip install dandi
-#
-# .. code-block:: python
-#
-#   from dandi.dandiapi import DandiAPIClient
-#
-#   dandiset_id = '000006'  # ephys dataset from the Svoboda Lab
-#   filepath = 'sub-anm372795/sub-anm372795_ses-20170718.nwb'  # 450 kB file
-#   with DandiAPIClient() as client:
-#       asset = client.get_dandiset(dandiset_id, 'draft').get_asset_by_path(filepath)
-#       s3_path = asset.get_content_url(follow_redirects=1, strip_query=True)
-
-####################
-# Finally, instantiate a :py:class:`~pynwb.NWBHDF5IO` object with the S3 URL and specify the driver as "ros3". This
-# will download metadata about the file from the S3 bucket to memory. The values of datasets are accessed lazily,
-# just like when reading an NWB file stored locally. So, slicing into a dataset will require additional time to
-# download the sliced data (and only the sliced data) to memory.
-#
-# .. code-block:: python
-#
-#   from pynwb import NWBHDF5IO
-#
-#   with NWBHDF5IO(s3_path, mode='r', load_namespaces=True, driver='ros3') as io:
-#       nwbfile = io.read()
-#       print(nwbfile)
-#       print(nwbfile.acquisition['lick_times'].time_series['lick_left_times'].data[:])
-
-####################
-# Method 2: s3fs
-# ~~~~~~~~~~~~~~
-# s3fs is a library that creates a virtual filesystem for an S3 store. With this approach, a virtual file is created
-# for the file and virtual filesystem layer will take care of requesting data from the s3 bucket whenever data is
-# read from the virtual file.
-#
-# First install s3fs:
-#
-# .. code-block:: bash
-#
-#   pip install s3fs
-#
-# Then in Python:
-#
-# .. code-block:: python
-#
-#   import s3fs
-#   import pynwb
-#   import h5py
-#
-#   fs = s3fs.S3FileSystem(anon=True)
-#
-#   f = fs.open("s3://dandiarchive/blobs/43b/f3a/43bf3a81-4a0b-433f-b471-1f10303f9d35", 'rb')
-#   file = h5py.File(f)
-#   io = pynwb.NWBHDF5IO(file=file, load_namespaces=True)
-#
-#   io.read()
```

Comment:
Since the ``s3fs`` builds on ``fsspec`` and since the ``s3fs`` code is simpler than using the full ``fsspec`` I think it would be nice to add a subsection to the ``Streaming Method 2: fsspec`` section that shows this code example for ``s3fs`` as well. In this way you can still keep the tutorial to the two main options but still mention the``s3fs`` option, which may be sufficient  and simpler for many users.

Reply:
So s3fs is a specific implementation of the fsspec protocol. I changed over to using the more generic fsspec library with http for three reasons:
1. This way we can use the http path as input, which is the same path that the ros3 driver takes, so it is clear how the two methods relate to each other.
2. Showing it this way makes it pretty clear how you would extend this to other backends.
3. This method also shows how to use caching, which I think is an important feature that a lot of users will want to leverage.
Do you really think this workflow is more complex than the s3fs syntax?

Reply:
I think the best would be if someone checks for what s3fs might be providing "on top" or in addition to pure straight fsspec. E.g. it could be some better choices of blocksize etc. I would have at least compared performance.

Reply:
is it true that s3fs provides additional convenience methods for interacting with S3 more specifically?

s3fs implements in [`S3FileSystem`](https://s3fs.readthedocs.io/en/latest/api.html#s3fs.core.S3FileSystem) which inherits from fsspec's `AsyncFileSystem` abstract class. My impression is that it just implements the method of a `fsspec.FileSystem` without much else besides maybe credential controls that we don't use. You can switch to an s3 backend by simply changing `fs=fsspec.filesystem("http")` to `fs=fsspec.filesystem("s3")`, but then you'll have to supply the s3 path, not the http path.

Reply:
It's hard to tell from the documentation what the specific differences are, but my impression is that s3fs adds support for ``async``, improved logging, and S3 specific implementation of the file system functions. Purely to read an NWB file, this may not be big difference. Pending further investigation, maybe what we could do is to add a note to indicate that there are a growing number of libraries build on ffspec that provide functionality specific for different file systems (e.g., S3, Google Cloud, or Azure) and link to https://filesystem-spec.readthedocs.io/en/latest/api.html?highlight=S3#other-known-implementations. We could then post alternate examples, such as the code using S3FS, as a GitHub Gist online and link to them from the same note. This would keep the docs clean and at the same time provide examples for folks that may want to use the more specific libraries build on fsspec. Given that we don't know how all the various options perform, we could also add a "disclaimer" along those lines.